errai-navigation/src/main/java/org/jboss/errai/ui/nav/client/local/Navigation.java

```diff
@@ -444,49 +448,71 @@ public void onAttachOrDetach(final AttachEvent event) {
 
+        accessEvent.fireAsync(component, new LifecycleCallback() {
           @Override
           public void callback(final boolean success) {
             if (success) {
-              locked = true;
+              NavigationControl showControl = new NavigationControl(Navigation.this, new Runnable() {
+                @Override
+                public void run() {
+                  try {
+                    // Fire IOC lifecycle event to indicate that the state of the
+                    // bean has changed.
+                    // TODO make this smarter and only fire state change event when
+                    // fields actually changed.
+                    stateChangeEvent.fireAsync(component);
+
+                    setCurrentPage(request.pageNode);
+                    currentWidget = componentWidget;
+                    currentComponent = component;
+                    navigatingContainer.setWidget(componentWidget);
+                    request.pageNode.pageShown(component, request.state);
+                  } finally {
+                    locked = false;
+                  }
+
+                  handleQueuedRequests(request, fireEvent);
+                }
+              }, new Runnable() {
+                @Override
+                public void run() {
+                  locked = false;
+                }
+              });
+
               try {
+                locked = true;
                 hideCurrentPage();
-                request.pageNode.pageShowing(component, request.state);
-
-                // Fire IOC lifecycle event to indicate that the state of the
-                // bean has changed.
-                // TODO make this smarter and only fire state change event when
-                // fields actually changed.
-                stateChangeEvent.fireAsync(component);
-
-                setCurrentPage(request.pageNode);
-                currentWidget = componentWidget;
-                currentComponent = component;
-                navigatingContainer.setWidget(componentWidget);
-                request.pageNode.pageShown(component, request.state);
-              } finally {
+                request.pageNode.pageShowing(component, request.state, showControl);
+              }
+              catch (Exception ex) {
                 locked = false;
+                throw ex;
               }
-
-              handleQueuedRequests(request, fireEvent);
-            }
-            else {
+            } else {
               request.pageNode.destroy(component);
             }
           }
         });
       }
     });
 
-    if (currentPage != null && currentWidget != null && currentComponent != null && currentWidget.asWidget() == navigatingContainer.getWidget()) {
-      currentPage.pageHiding(Factory.maybeUnwrapProxy(currentComponent), control);
-    }
-    else {
+    if (!hiding && currentPage != null && currentWidget != null && currentComponent != null && currentWidget.asWidget() == navigatingContainer.getWidget()) {
```

Comment:
I'm not sure I understand the purpose of this `hiding` in this condition. If I am interpreting this change correctly, I think this will introduce problems if you try to navigate away from a page with a `@PageHiding` method having a `NavigationControl`.

For example, say I have a page with a form on it and the page uses `@PageHiding` to block navigation away from the page while the form has not been submitted. With this change if a user tries to navigate a way, the first time it will be blocked by `@PageHiding`, but the second time the `@PageHiding` method won't be called and navigation will just proceed.